### PR TITLE
Added Cloud Anchor support to ArNode and sample application

### PR DESCRIPTION
**Pull request by grassydragon**
Why have we decided to support Cloud Anchors right in the `ArNode`:

- The required code is relatively small
- The `ArNode` already detaches the anchor in the `destroy` method and, therefore, cancel the resolve request
- Since all AR nodes are inherited from the `ArNode` it makes it possible to use Cloud Anchors when a model is placed using the `ArModelNode` and then shared or when the base `ArNode` is used directly or inherited to support more advanced features

Currently we are testing the new methods and improving the sample application.

### CODE REVIEW

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 1]
```diff
@@ -65,6 +66,12 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
             onAnchorChanged(value)
         }
 
+    /**
+     * ### Whether a Cloud Anchor is currently being hosted or resolved
+     */
+    var requestInProgress = false
+        private set
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:73]:
```suggestion
val cloudAnchorState : CloudAnchorState get() = anchor?.cloudAnchorState ?: CloudAnchorState.NONE
/**
 * ### Whether a Cloud Anchor is currently being hosted or resolved
 */
var cloudAnchorTaskInProgress = false
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 2]
```diff
@@ -83,9 +92,22 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+
+        if (requestInProgress) {
+            val state = anchor.cloudAnchorState
+
+            // Call the listener when the request completes successfully or with an error
+            if (state != CloudAnchorState.NONE && state != CloudAnchorState.TASK_IN_PROGRESS) {
+                requestInProgress = false
+                onRequestComplete?.invoke(anchor, state == CloudAnchorState.SUCCESS)
+                onRequestComplete = null
+            }
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:110]: Use `cloudAnchorState` directly

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 3]
```diff
@@ -75,6 +82,8 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
      */
     var onAnchorChanged: ((node: Node, anchor: Anchor?) -> Unit)? = null
 
+    private var onRequestComplete: ((anchor: Anchor, success: Boolean) -> Unit)? = null
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:85]:
```suggestion
private var onCloudAnchorTaskCompleted: ((anchor: Anchor, success: Boolean) -> Unit)? = null
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 4]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
 
+    /**
+     * ### Hosts a Cloud Anchor based on the [anchor]
+     *
+     * The [anchor] is replaced with a new anchor returned by [Session.hostCloudAnchorWithTtl].
+     *
+     * @param ttlDays The lifetime of the anchor in days. See [Session.hostCloudAnchorWithTtl] for more details.
+     * @param onRequestComplete Called when the request completes successfully or with an error.
+     */
+    fun hostCloudAnchor(ttlDays: Int = 1, onRequestComplete: (anchor: Anchor, success: Boolean) -> Unit) {
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:178]:
```suggestion
fun hostCloudAnchor(ttlDays: Int = 1, onTaskCompleted: (anchor: Anchor, success: Boolean) -> Unit) {
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 5]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+        checkRequestState()
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:179]:
```suggestion
if (cloudAnchorTaskInProgress) throw IllegalStateException("A Cloud Anchor request is already in progress")
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 6]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+
+        if (anchor == null) throw IllegalStateException("The anchor shouldn't be null")
+
+        anchor = arSession?.hostCloudAnchorWithTtl(anchor, ttlDays)
+        requestInProgress = true
+        this.onRequestComplete = onRequestComplete
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:185]:
```suggestion
onCloudAnchorTaskCompleted = onTaskCompleted
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 7]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+
+    /**
+     * ### Resolves a Cloud Anchor
+     *
+     * The [anchor] is replaced with a new anchor returned by [Session.resolveCloudAnchor].
+     *
+     * @param cloudAnchorId The Cloud Anchor ID of the Cloud Anchor.
+     * @param onRequestComplete Called when the request completes successfully or with an error.
+     */
+    fun resolveCloudAnchor(cloudAnchorId: String, onRequestComplete: (anchor: Anchor, success: Boolean) -> Unit) {
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:196]:
```suggestion
fun resolveCloudAnchor(cloudAnchorId: String, onTaskCompleted: (anchor: Anchor, success: Boolean) -> Unit) {
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 8]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+        checkRequestState()
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:197]:
```suggestion
if (cloudAnchorTaskInProgress) throw IllegalStateException("A Cloud Anchor request is already in progress")
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 9]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+
+        anchor = arSession?.resolveCloudAnchor(cloudAnchorId)
+        requestInProgress = true
+        this.onRequestComplete = onRequestComplete
+    }
+
+    private fun checkRequestState() {
+        if (requestInProgress) throw IllegalStateException("The request is already in progress")
+    }
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:206]:
```suggestion
```

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt [hunk 10]
```diff
@@ -145,6 +167,57 @@ open class ArNode() : ModelNode(), ArSceneLifecycleObserver {
...
+
+    /**
+     * ### Cancels a resolve request
+     *
+     * The [anchor] is detached to cancel the resolve request.
+     */
+    fun cancelResolveRequest() {
```
Comment (ThomasGorisse) [inline arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:213]:
```suggestion
cancelCloudAnchorTask() {
```

Comment (ThomasGorisse) [reply arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:213]: You must also add the call to it inside the `onDestroy()`

**Review by ThomasGorisse — CHANGES_REQUESTED** [review]
Great. Mostly some wording modifactions and we can merge it.

File: arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt (hunk 10 above)
Comment (grassydragon) [reply arsceneview/src/main/java/io/github/sceneview/ar/node/ArNode.kt:213]: According to the documentation it's only possible to cancel a resolve request this way: https://developers.google.com/ar/develop/java/cloud-anchors/developer-guide#cancel_or_remove_a_cloud_anchor

**Comment by grassydragon** [comment]
I've decided to add a "Reset" instead of an "Exit" button to resolve the Cloud Anchor without exiting the application.

**Pull request merged**